src/mlpack/methods/decision_tree/gini_gain.hpp

```diff
@@ -84,28 +85,60 @@ class GiniGain
     if (UseWeights)
     {
       // Sum all the weights up.
-      double accWeights[4] = { 0.0, 0.0, 0.0, 0.0 };
-
+      // double accWeights[4] = { 0.0, 0.0, 0.0, 0.0 };
+      double accWeights = 0.0;
       // SIMD loop: add counts for four elements simultaneously (if the compiler
       // manages to vectorize the loop).
-      for (size_t i = 3; i < labels.n_elem; i += 4)
-      {
-        const double weight1 = weights[i - 3];
-        const double weight2 = weights[i - 2];
-        const double weight3 = weights[i - 1];
-        const double weight4 = weights[i];
+      #pragma omp parallel for reduction (+:accWeights)
 
-        counts[labels[i - 3]] += weight1;
-        counts2[labels[i - 2]] += weight2;
-        counts3[labels[i - 1]] += weight3;
-        counts4[labels[i]] += weight4;
-
-        accWeights[0] += weight1;
-        accWeights[1] += weight2;
-        accWeights[2] += weight3;
-        accWeights[3] += weight4;
```

Comment:
Note that this loop was specifically written to get the compiler to produce SIMD instructions that will add together four elements in one instruction; that's why it appears to be unrolled so that four elements are handled at once.

If you want to use OpenMP with this code, I think that could be nice, but the way to do it will need to preserve the SIMD instruction structure.  Consider, e.g., if there were two cores, then split `labels` in half, and have the first core do the first half with SIMD instructions, and the second do the second half with SIMD instructions.

I'd strongly consider looking at the assembly output of the loop here to make sure that when you are done, that the compiler is actually using AVX2 SIMD instructions for the addition. :+1:

Reply:
@rcurtin 

Thank you for your review.

Correct me if I am wrong.
gini_gain's original version uses SIMD instructions(instruction level parallelism) for better performance. This means that compiler executes 4 instructions in 1 cycle.
The updated version uses for construct of OpenMP(task parallelism) for better performance. This means that compiler creates threads assigns them independent tasks to do.
![Screenshot from 2020-02-18 10-59-13](https://user-images.githubusercontent.com/36797859/74706874-19a64b00-523e-11ea-81c0-571be1edf558.png)

As I understand you want me to use SIMD along with OpenMP to increase performance.
But I dont think that is a great idea because then it will have a part of code (i.e for counts) which will we will have to execute in critical region of the program.
(i.e almost 1/2 the code in critical region of the program). This would kill its performance.

Note:- for construct of OpenMP does what you said. It creates threads and every thread executes its own block of the program. So in this version of gini_gain we will have have accWeights being executed in parallel and 1 instruction in critical section (counts) which I think will perform better than using SIMD and OpenMP together.